pkg/services/pluginsintegration/renderer/renderer.go

```diff
@@ -0,0 +1,125 @@
+package renderer
+
+import (
+	"context"
+	"errors"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/plugins"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin/pluginextensionv2"
+	"github.com/grafana/grafana/pkg/plugins/backendplugin/provider"
+	"github.com/grafana/grafana/pkg/plugins/config"
+	"github.com/grafana/grafana/pkg/plugins/envvars"
+	"github.com/grafana/grafana/pkg/plugins/manager/loader"
+	"github.com/grafana/grafana/pkg/plugins/manager/pipeline/bootstrap"
+	"github.com/grafana/grafana/pkg/plugins/manager/pipeline/discovery"
+	"github.com/grafana/grafana/pkg/plugins/manager/pipeline/initialization"
+	"github.com/grafana/grafana/pkg/plugins/manager/pipeline/termination"
+	"github.com/grafana/grafana/pkg/plugins/manager/pipeline/validation"
+	"github.com/grafana/grafana/pkg/plugins/manager/registry"
+	"github.com/grafana/grafana/pkg/plugins/manager/signature"
+	"github.com/grafana/grafana/pkg/plugins/manager/sources"
+	"github.com/grafana/grafana/pkg/services/rendering"
+)
+
+func ProvideService(cfg *config.Cfg, registry registry.Service, licensing plugins.Licensing) (*Manager, error) {
+	l, err := createLoader(cfg, registry, licensing)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Manager{
+		cfg:    cfg,
+		loader: l,
+		log:    log.New("plugins.renderer"),
+	}, nil
+}
+
+type Manager struct {
+	cfg    *config.Cfg
+	loader loader.Service
+	log    log.Logger
+
+	renderer *Plugin
+}
+
+type Plugin struct {
+	plugin *plugins.Plugin
+
+	started bool
+}
+
+func (p *Plugin) Client() (pluginextensionv2.RendererPlugin, error) {
+	if !p.started {
+		return nil, errors.New("renderer plugin not started")
+	}
+
+	if p.plugin.Renderer == nil {
+		return nil, errors.New("renderer client not available")
+	}
+
+	return p.plugin.Renderer, nil
+}
+
+func (p *Plugin) Start(ctx context.Context) error {
+	p.started = true
+	return p.plugin.Start(ctx)
+}
+
+func (p *Plugin) Version() string {
+	return p.plugin.JSONData.Info.Version
+}
+
+func (m *Manager) Renderer(ctx context.Context) (rendering.Plugin, bool) {
+	if m.renderer != nil {
+		return m.renderer, true
+	}
+
+	ps, err := m.loader.Load(ctx, sources.NewLocalSource(plugins.ClassExternal, []string{m.cfg.PluginsPath}))
+	if err != nil {
+		m.log.Error("Failed to load renderer plugin", "error", err)
+		return nil, false
+	}
+
+	if len(ps) == 1 {
+		m.renderer = &Plugin{plugin: ps[0]}
+		return m.renderer, true
+	}
+
+	return nil, false
+}
+
+func createLoader(cfg *config.Cfg, pr registry.Service, l plugins.Licensing) (loader.Service, error) {
```

Comment:
I guess this is the con, having a whole loader just for one "plugin"

Reply:
I think this is okay actually. In fact I was originally hoping that the pipeline changes would be capable for this. Considering the mixing of logic we still have in places such as Core vs External for example, they could also be good opportunities - but I do think we'd need to organize it in a way that feels more natural.

Reply:
I think it looks good being able to replace steps in the pipeline like this! However we probably need some helper method to return a "default" pipeline (with sane steps), and allow each loader to customize or replace the steps as needed.

Maybe something like this:

```go
func createLoader(cfg *config.Cfg, pr registry.Service, l plugins.Licensing) (loader.Service, error) {
	l, err := loader.NewDefaultLoader(
		loader.WithDiscovery(
			disovery.NewDefault(
				discovery.WithPermittedPluginTypes(plugins.TypeRenderer),
			)
		),
		loader.WithValidation(validation.WithUnsigned()),
		loader.WithInitialization(
			initialization.WithProvider(provider.RendererProvider),
			initialization.WithRegistry(pr)
		),
		loader.WithTermination(termination.WithRegistry(pr))
	)
	if err != nil {
		return nil, fmt.Errorf("default loader: %w", err)
	}
	return l, err
}
```

We need to think a bit about the API and make sure it allows to replace/extend things without too much hassle.

But I think that's a refactoring step for another day and I'm personally happy to keep this piece of code as-is for the time being!

Reply:
Yeah that looks awesome @xnyo 😍 Would definitely like to see us invest more time into something like that if we find we start to use this pattern more. Thanks!